src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,664 @@
+     */
+    T orElseSet(Supplier<? extends T> supplier);
+
+    // Convenience methods
+
+    /**
+     * Sets the content to the provided {@code value}, or, if already set,
+     * throws {@code IllegalStateException}.
+     * <p>
+     * When this method returns (or throws an exception), the content is always set.
+     *
+     * @param value to set
+     * @throws IllegalStateException if the content was already set
+     */
+    void setOrThrow(T value);
+
+    // Object methods
+
+    /**
+     * {@return {@code true} if {@code this == obj}, {@code false} otherwise}
+     *
+     * @param obj to check for equality
+     */
+    boolean equals(Object obj);
+
+    /**
+     * {@return the {@linkplain System#identityHashCode(Object) identity hash code} of
+     *          {@code this} object}
+     */
+    int hashCode();
+
+    // Factories
+
+    /**
+     * {@return a new unset stable value}
+     * <p>
+     * An unset stable value has no content.
+     * <p>
+     * The returned stable value is not {@link Serializable}.
+     *
+     * @param <T> type of the content
+     */
+    static <T> StableValue<T> of() {
+        return StableValueImpl.of();
+    }
+
+    /**
+     * {@return a new pre-set stable value with the provided {@code content}}
+     * <p>
+     * The returned stable value is not {@link Serializable}.
+     *
+     * @param content to set
+     * @param <T>     type of the content
+     */
+    static <T> StableValue<T> of(T content) {
+        final StableValue<T> stableValue = StableValue.of();
+        stableValue.trySet(content);
+        return stableValue;
+    }
+
+    /**
+     * {@return a new stable supplier}
+     * <p>
+     * The returned {@linkplain Supplier supplier} is a caching supplier that records
+     * the value of the provided {@code original} supplier upon being first accessed via
+     * the returned supplier's {@linkplain Supplier#get() get()} method.
+     * <p>
+     * The provided {@code original} supplier is guaranteed to be successfully invoked
+     * at most once even in a multi-threaded environment. Competing threads invoking the
+     * returned supplier's {@linkplain Supplier#get() get()} method when a value is
+     * already under computation will block until a value is computed or an exception is
+     * thrown by the computing thread.
+     * <p>
+     * If the provided {@code original} supplier throws an exception, it is relayed
+     * to the initial caller and no content is recorded.
+     * <p>
+     * The returned supplier is not {@link Serializable}.
+     * <p>
+     * If the provided {@code original} supplier recursively calls the returned
+     * supplier, an {@linkplain IllegalStateException} will be thrown.
+     *
+     * @param original supplier used to compute a cached value
+     * @param <T>      the type of results supplied by the returned supplier
+     */
+    static <T> Supplier<T> supplier(Supplier<? extends T> original) {
```

Comment:
I think that here and in all other functional/collection factories, we should document that `toString` will never cause a value to be computed.